backends/dpdk/dpdkArch.cpp

```diff
@@ -467,6 +467,215 @@ bool CollectMetadataHeaderInfo::preorder(const IR::Type_Struct *s) {
     return true;
 }
 
+/* This function processes each header structure and align each non-aligned header
+   into 8-bit aligned header.[Restricted to only continuos non-aligned header fields]
+   For example :
+       header ipv4_t {
+            bit<4>  version;
+            bit<4>  ihl;
+            bit<8>  diffserv;
+            bit<32> totalLen;
+            bit<16> identification;
+            bit<3>  flags;
+            bit<13> fragOffset;
+            bit<8>  ttl;
+            bit<8>  protocol;
+            bit<16> hdrChecksum;
+            bit<32> srcAddr;
+            bit<32> dstAddr;
+       }
+       is converted into
+       header ipv4_t {
+            bit<8>  version_ihl;      // "version" and "ihl" combined for 8-bit alignment
+                                         resulting into 8 bit "version_ihl" field
+            bit<8>  diffserv;
+            bit<32> totalLen;
+            bit<16> identification;
+            bit<16> flags_fragOffset; // "flag" and "fragOffset" combined for 8-bit aligment
+                                         resulting into 16-bit "flags_fragOffset" field
+            bit<8>  ttl;
+            bit<8>  protocol;
+            bit<16> hdrChecksum;
+            bit<32> srcAddr;
+            bit<32> dstAddr;
+       }
+*/
+const IR::Node *AlignHdrMetaField::preorder(IR::Type_StructLike *st) {
```

Comment:
you are missing the processing of StructExpression, which can be used to initialize headers. These have to be rewritten too.

Reply:
I have added few new testcase for StructExpression usage and its alignment. Its working. Please check the testcase. Also please suggest for the usage/testcase for StructExpression

Reply:
I guess you don't need to handle StructExpression because the backend runs the CopyStructures pass which disassembles such expressions into individual field operations. However, if you decide *not* to run that pass, you will have to modify this pass to handle StructExpressions.

Reply:
Yes, We are going to run the pass the flatten the StructExpression into individual field. So no more handling is required for StructExpressions